crates/wasmtime/src/store.rs

```diff
@@ -145,6 +185,44 @@ impl Store {
             .map(|x| x as _)
     }
 
+    /// Calls the given closure with the current context value of the given type.
+    ///
+    /// If a context value of the given type is not present, the closure is passed `None`.
+    ///
+    /// Returns the value returned by the closure.
+    pub fn with_context<T: Any, F, R>(&self, func: F) -> R
+    where
+        F: Fn(Option<&mut T>) -> R,
+    {
+        let mut values = self.inner.context_values.borrow_mut();
```

Comment:
Hm we may wish to brainstorm a different API for this. I'm worried about this causing unexpected panics in applications because you couldn't, for example, call this method recursively twice. I fear that would hamper usage of this API quite a lot. For example if you're implementing a WASI API which calls back into the module to malloc you'd need to be extremely careful to not `with_context` while malloc is being run.

Reply:
I think the ideal API here would be `fn get<T>(&self) -> Option<&T>` perhaps.

Could we not support `remove` and if consumers want it then they could have `T = RefCell<Option<U>>` perhaps?

Reply:
+1 for exposing `&T` instead of `&mut T`

if users are the ones adding ref cells, then they are aware of the re-entrancy concerns and can design around that, rather than being surprised when they finally hit a corner case and have already designed and implemented a bunch of stuff that now needs to be redone

Reply:
> Could we not support `remove` and if consumers want it then they could have `T = RefCell<Option<U>>` perhaps?

Yeah, good point. I guess switching to `&T` alone isn't enough, we also have to not support removal to make reentrancy panic-free.

Reply:
Worth documenting for users how they would get mutability and the ability to remove context via `RefCell<Option<T>>` in the doc comments.

Reply:
That's a good point about reentrancy and moving the `RefCell` into the context value would give the callers more control over mutation scope.

This will necessitate a method on `Wasi` or `WasiCtx` to set the context in the store with the exact type the generated `Wasi` host functions expect, though (it's how I had it in the RFC, but this was a poor attempt at a "cleaner" API).

I'll get this fixed.